### ISSUE TITLE
parallel distributed lazy for big(er) meshes

### ISSUE THREAD

**Issue by anderson2981** [illinois-ceesd/mirgecom#599]
Parallel distributed lazy runs out of memory during compilation on a 3M element mesh, 3D, second order.

Node counts of 1, 2, 4, 8 were tried on lassen. With 4 ranks (gpus) utilized on each node.

This mesh will run on 4 nodes in eager mode.

The data sets can be found in the distributed lazy branch of the isolator driver

[isolator](https://github.com/illinois-ceesd/drivers_y2-isolator/tree/distributed-lazy)

in the test_mesh subdirectory. halfX is linked to the 3M element mesh, eigthX and quarterX are smaller meshes which run successfully. There is also a 21M element mesh located in the oneX directory. 

The meshes are not checked into the repo, but can be generated with the make_mesh.sh scripts located in data/3D/<halfX>. These scripts require gmsh 4.9.3 to be installed.

**Comment by matthiasdiener** [comment]
I believe this is the error message?

```
  File "/p/gpfs1/manders/work/CEESD/Drivers/CEESD-Y2-isolator-parallel/emirge/pytato/pytato/distributed.py", line 660, in execute_distributed_partition
    exec_ready_part(partition.parts[pid])
  File "/p/gpfs1/manders/work/CEESD/Drivers/CEESD-Y2-isolator-parallel/emirge/pytato/pytato/distributed.py", line 615, in exec_ready_part
    _evt, result_dict = prg_per_partition[part.pid](queue, **inputs)
  File "/p/gpfs1/manders/work/CEESD/Drivers/CEESD-Y2-isolator-parallel/emirge/pytato/pytato/target/loopy/__init__.py", line 146, in __call__
    return self.program(queue,
  File "/p/gpfs1/manders/work/CEESD/Drivers/CEESD-Y2-isolator-parallel/emirge/loopy/loopy/translation_unit.py", line 342, in __call__
    return pex(*args, **kwargs)
  File "/p/gpfs1/manders/work/CEESD/Drivers/CEESD-Y2-isolator-parallel/emirge/loopy/loopy/target/pyopencl_execution.py", line 367, in __call__
    return translation_unit_info.invoker(
  File "/usr/WS1/xpacc/Users/manders/software/Install/Lassen/Conda/envs/mirgeDriver.Y2isolator-parallel/lib/python3.9/site-packages/pytools/py_codegen.py", line 150, in __call__
    return self.func(*args, **kwargs)
  File "<generated code for 'invoke__pt_kernel_loopy_kernel'>", line 248, in invoke__pt_kernel_loopy_kernel
  File "<generated code for 'invoke__pt_kernel_loopy_kernel'>", line 31, in _lpy_host__pt_kernel
pyopencl._cl.RuntimeError: clEnqueueNDRangeKernel failed: OUT_OF_RESOURCES
```

cc: @kaushikcfd @inducer

**Comment by anderson2981** [comment]
Yup, that's the one I usually get. 

I ran on 16 nodes last night and got this error message instead
`CUDA_ERROR_INVALID_PTX: a PTX JIT compilation failed`

**Comment by matthiasdiener** [comment]
Unfortunately, the com-bozzle case does not seem to reproduce this issue:

- Com-bozzle (mixalot.py), GPU, porter


|   | 1 rank   | 2 ranks  |
| -------- | -------- | -------- |
| Eager     | ❌   (OUT_OF_RESOURCES)    | ❌   (OUT_OF_RESOURCES)    |
| Lazy     |  ✅     | ✅    |

**Comment by MTCam** [comment]
> Unfortunately, the [com-bozzle](https://github.com/illinois-ceesd/drivers_bozzle/blob/state-handling/mixalot.py) case does not seem to reproduce this issue:

Agreed. For some reason Eager tops out at ~23k elements, but lazy is able to run all the way out to 444k elements without seeing this issue.  For a failing lazy case in `mixalot.py`, set xscale=yscale=zscale=16.

**Comment by matthiasdiener** [comment]
An interesting test case for `mixalot.py` is 

```
    x_scale = 8
    y_scale = 8
    z_scale = 16
```

which succeeds on one rank but fails with OOM on two ranks.

**Comment by anderson2981** [comment]
Running with the newest fusion array contractor, I get this error when compiling isolator_injection_run.py.

[error](https://gist.github.com/anderson2981/9027f5ba3483bd4c9ce0a4b74d60391e)

**Comment by anderson2981** [comment]
I think I fixed the issue, at least temporarily. I updated the instructions linked above to include descriptions for using the production scale meshes. Currently none of these meshes/runs are able to be complied with lazy. They fail with the error message 
```CUDA_ERROR_INVALID_PTX: a PTX JIT compilation failed```

**Comment by kaushikcfd** [comment]
`nspecies=7` is still an issue for those branches. I'll take a look at what could be done for those cases.

**Comment by matthiasdiener** [comment]
The following two branches should address the arg size issue on Nvidia:

- Loopy: https://github.com/matthiasdiener/loopy/tree/fusion_arg_reduction
- Meshmode: https://github.com/matthiasdiener/meshmode/tree/production-fusion

In particular, I was able to run `drivers_y2-isolator/smoke_test_injection_3d` with `nspecies=7` successfully on Nvidia GPUs with these branches.
